contract/src/main/java/io/pravega/schemaregistry/contract/data/VersionInfo.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.contract.data;
+
+import io.pravega.common.ObjectBuilder;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+/**
+ * Version information object that encapsulates properties that uniquely identify a specific version of a schema within a group.
+ * 
+ * {@link VersionInfo#type} is same as {@link SchemaInfo#type} which represents the object type for which the version is computed. 
+ * {@link VersionInfo#version} the registry assigned monotonically increasing version number for the schema for specific object type.
+ * Since the version number is per object type, so type and version number forms a unique pair. 
+ * {@link VersionInfo#ordinal} Absolute ordinal of the schema for all schemas in the group. This uniquely identifies the 
+ * version within a group. 
+ */
+@Data
+@Builder
+@AllArgsConstructor
+public class VersionInfo {
+    /**
+     * Object type which is declared in the corresponding {@link SchemaInfo#type} for the schemainfo that is identified 
+     * by this version info. 
+     */
+    private final String type;
+    /**
+     * A version number that identifies the position of schema among other schemas in the group that share the same 'type'.
+     */
+    private final int version;
+    /**
+     * A position identifier that uniquely identifies the schema within a group and represents the order in which this
+     * schema was included in the group. 
+     */
+    private final int ordinal;
```

Comment:
tkaitchuck 5 days ago Member
```
Are the definitions of ordinal and version switched? From the names, it sounds like version would be user defined and ordinal would be assigned by the service?
```
@shiveshr
shiveshr 5 days ago Author Collaborator
```
Both version and ordinal are assigned by the service. Ordinal is like a unique Id within a group that also helps identify the position.
Version is primarily for human readability. Schema for version 2 of type "User", schema for version 3 of type "Address" etc.
```
@tkaitchuck
tkaitchuck 3 days ago Member
```
Ok, so users are going to construct these, so what happens if they don't specify an ordinal or the ordinal and version don't match? It would seem a lot nicer from an API point of view if the user didn't have to specify the non-human readable number.
```
@shiveshr
shiveshr 3 days ago Author Collaborator
```
Users do not have the information to construct this object themselves.
It is generated by the service and given as a response to user queries.
```

Basically from REST api exposes 2 sets of apis in the service to get/delete a specific schema version --
1. use `schemaType` + `version` 
2. `versionOrdinal`
Both these uniquely identify the same schema in the group and work on the same object and function in the same way. 
The versionInfo object returned by the service has all three information.

Reply:
If the user can't create these, then they should really be combined into a single long so it is opaque to the caller.

Reply:
It provides some information to the user though. 
Like the version number tells them which specific schema is placed among the different versions of the schema. 
And the ordinal gives them a unique identifier to identify this schema version within the group. 
Our pravega applications tag the data with "encoding id". But as a hypothetical, if someone were to tag the data with only the version id, they could choose to use the ordinal. 
So this object is not opaque and provides these information to the users.

Reply:
> Like the version number tells them where specific schema is placed among the different versions of the schema.

A user can't see into our service. The only information they have is via the API. So "where it is placed" is an implementation detail that they cannot see and should not be aware. 

> Our pravega applications tag the data with "encoding id". But as a hypothetical, if someone were to tag the data with only the version id, they could choose to use the ordinal.

I don't see how this would work. What API takes the ordinal and not the version? Also what is the advantage? EncodingId is 4 bytes and so is ordinal. Are they the same value? Can they be made into the same value?

> So this object is not opaque and provides these information to the users.

I don't understand what information in provides. Which Javadoc should I read?

Reply:
> A user can't see into our service. The only information they have is via the API. So "where it is placed" is an implementation detail that they cannot see and should not be aware.

"Where" = where in the context of schemas they have added.. Schemas are assigned version number in the order in which they are added - first schema is version 0. second schema is version 1. so on and so forth... Users would obviously want to know that information when looking at a schema. 

> I don't see how this would work. What API takes the ordinal and not the version? Also what is the advantage? EncodingId is 4 bytes and so is ordinal. Are they the same value? Can they be made into the same value?

What will not work? 
Firstly, there are two 4 byte ids that the service generates to represent two different concepts. 
1. version ordinal ==> identifies a schema version within a group.
2. encoding id ==> identifies a pair of schema version and codec type. 

The serializers we have created for Pravega applications make use of encoding id and allow for codecs. 
Someone could use schema registry without caring about codecs and would only care about the version id for the schema to tag their data with.

Reply:
The user does not care about these distinctions. The only thing that matters is what can be done with the value. From a user point of view:
Given a (version + type + group) they can get the schema and any related information.
Given a (ordinal + group) they can get the schema and any related information.
Given a (encodingId) they can get the schema and any related information.

But, they are all 4 bytes. So `encodingId` is strictly better than `ordinal` which is strictly better than `versionId`.

Reply:
> But, they are all 4 bytes. So encodingId is strictly better than ordinal which is strictly better than versionId.

Yes that is partially correct, however, encodingId and ordinal are not interchangable.

VersionId alone is not sufficient and needs type to uniquely identify schema version. 
It is interchangable with ordinal identifies a schema version uniquely.

encodingId identifies a pair of codec and schema version. Whose purpose is different from a version alone. 
A version tells the order in which schema evolution happened and identifies a specific schema in the group.

encoding id tells about how a data is encoded - codecType and version.

Reply:
Renamed ordinal to schemaId.